nixos/modules/services/web-apps/rutorrent.nix

```diff
@@ -0,0 +1,281 @@
+                @define('LOG_RPC_FAULTS', true, true);
+
+                // for php
+                @define('PHP_USE_GZIP', false, true);
+                @define('PHP_GZIP_LEVEL', 2, true);
+
+                $schedule_rand = 10;			// rand for schedulers start, +0..X seconds
+
+                $do_diagnostic = true;
+                $log_file = '${cfg.home}/logs/errors.log';		// path to log file (comment or leave blank to disable logging)
+
+                $saveUploadedTorrents = true;		// Save uploaded torrents to profile/torrents directory or not
+                $overwriteUploadedTorrents = false;     // Overwrite existing uploaded torrents in profile/torrents directory or make unique name
+
+                $topDirectory = '/';			// Upper available directory. Absolute path with trail slash.
+                $forbidUserSettings = false;
+
+                $scgi_port = 0;
+                $scgi_host = "unix://${cfg.rpcSocket}";
+
+                $XMLRPCMountPoint = "/RPC2";		// DO NOT DELETE THIS LINE!!! DO NOT COMMENT THIS LINE!!!
+
+                $pathToExternals = array(
+                  "php" 	=> ''',			// Something like /usr/bin/php. If empty, will be found in PATH.
+                  "curl"	=> ''',			// Something like /usr/bin/curl. If empty, will be found in PATH.
+                  "gzip"	=> ''',			// Something like /usr/bin/gzip. If empty, will be found in PATH.
+                  "id"	=> ''',			// Something like /usr/bin/id. If empty, will be found in PATH.
+                  "stat"	=> ''',			// Something like /usr/bin/stat. If empty, will be found in PATH.
+                );
+
+                $localhosts = array( 			// list of local interfaces
+                  "127.0.0.1",
+                  "localhost",
+                );
+
+                $profilePath = '../share';		// Path to user profiles
+                $profileMask = 0777;			// Mask for files and directory creation in user profiles.
+                // Both Webserver and rtorrent users must have read-write access to it.
+                // For example, if Webserver and rtorrent users are in the same group then the value may be 0770.
+
+                $tempDirectory = null;			// Temp directory. Absolute path with trail slash. If null, then autodetect will be used.
+
+                $canUseXSendFile = false;		// If true then use X-Sendfile feature if it exist
+
+                $locale = "UTF8";
+            '';
+          in {
+            wantedBy = [ "multi-user.target" ];
+            before = [ "phpfpm-rutorrent.service" ];
+            script = ''
+              ln -sf ${pkgs.rutorrent}/{css,images,js,lang,index.html} ${cfg.home}/
+              mkdir -p ${cfg.home}/{conf,logs,plugins} ${cfg.home}/share/{settings,torrents,users}
+              ln -sf ${pkgs.rutorrent}/conf/{access.ini,plugins.ini} ${cfg.home}/conf/
+              ln -sf ${rutorrentConfig} ${cfg.home}/conf/config.php
+
+              cp -r ${pkgs.rutorrent}/php ${cfg.home}/
```

Comment:
I have to copy instead of symlink the php files because it tries to import files from the `share` directory using a relative path. Using symlink it search relative to the nix store path instead of the `${cfg.home}`.

I find better to use symlink and tried to find a way to configure php so it would import files relative to the symlink and not it's target. I didn't find anything, is it possible? Otherwise I might end up patching the relative paths.

Reply:
Same things for the plugins.

Reply:
Maybe we could do this with `BindReadOnlyPaths=` in the systemd service? 

For example:
```
BindReadOnlyPaths = [ "${rutorrentConfig}:${cfg.home}/conf/config.php" ];
```

Reply:
@iv-nn So I came up with a bit of a magical solution using bind mounts. This is probably even more hacky, but I kind of like it because it leaves the `/var/lib/rutorrent` clear of non-state clutter.

Basiaclly what I do is I set the nginx `root = pkgs.rutorrent`, so it points to the ruTorrent nix store directly. Then I mount the `config.php` file on top of it, along with the whitelisted plugins. I also set the `$profilePath = '${cfg.home}';` so it points directly to the data dir. I do all of this in the phpfpm service file, so we don't even need the setup service.

Here's the modified `phpfpm-rutorrent` service:
```
phpfpm-rutorrent.serviceConfig = {
  ExecStartPre = [
    (pkgs.writeShellScript "setup-rutorrent" ''
      mkdir -p ${cfg.home}/{logs,settings,torrents,users}
      chown -R rutorrent:rutorrent ${cfg.home}/{logs,settings,torrents,users}
      chmod -R 755 ${cfg.home}/{logs,settings,torrents,users}
    '')
  ];
  TemporaryFileSystem = [ "${pkgs.rutorrent}/plugins" ];
  BindReadOnlyPaths = [
    "${rutorrentConfig}:${pkgs.rutorrent}/conf/config.php"
  ] ++ map (p: "${pkgs.rutorrent}/plugins/${p}") cfg.plugins;
};
```

What do you think?

Reply:
I like the idea of using bind mounts, it's clever. But I think I would rather mount the store files in the `dataDir` than the other way around.

But seeing your other comment I realise changing `$profilePath` should be enough.

Reply:
The plugins dir is the only issue really. With the modified config file you could easily have nginx point directly to `/nix/store/...` and load the settings through an env var, without any need for any bind mounts or symlinking.

Tbh, I would just skip the whole NixOS plugin management because you can manage plugins in the UI anyway. And it would greatly simplify the whole setup, though at the cost of unnecessary dependencies. Up to you though.

Reply:
I kept the copy for now because I thought maybe it's an issue to tie the plugin setup to the php-fpm service. Shouldn't there be a way to use this service without setting `nginx.enable = true`, using some other web server?

Does using copy pose a real problem?

It even allow to use custom plugins, although a better way would be to allow adding them directly through the `plugins` option.

Reply:
@iv-nn I think it's fine, forget it. I've seen Nix services do it both ways. Usually using the `/nix/store/...` as the vhost root when it's simple enough, or by copying the application into the state directory in more complex cases. I think with plugins it might be complex enough to keep it as it is. Regardless, `nginx.enable = true` is not necessary as you can always point your custom vhost `root` to `pkgs.rutorrent`.